### PR TITLE
Fix crash in ref-returning property

### LINKED ISSUE #44153: Crash on 'ref ErrorType' property
**Version Used**: 
b1913b8

**Steps to Reproduce**:

```cs
        [Fact]
        public void RefErrorPropertyCrash()
        {
            CompileAndVerify(@"
public class C {
    public ref ERROR Prop => throw null!;
}
").VerifyDiagnostics();
        }
```

~~This is a regression from 3.6.0-4.20228.5 (9bc8f80f)~~ Actually checking out this commit and running the test **does** cause the crash. For some reason the crash is exhibited when compiling this program in some contexts but not others. It does crash VS when pasted into a C# file.

Stack:
```
    System.NullReferenceException : Object reference not set to an instance of an object.
    Stack Trace:
       at Microsoft.CodeAnalysis.CSharp.DefiniteAssignmentPass.MarkFieldsUsed(TypeSymbol type) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\DefiniteAssignment.cs:line 1941
   at Microsoft.CodeAnalysis.CSharp.DefiniteAssignmentPass.WriteArgument(BoundExpression arg, RefKind refKind, MethodSymbol method) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\DefiniteAssignment.cs:line 1900
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitReturnStatementNoAdjust(BoundReturnStatement node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 1744
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitReturnStatement(BoundReturnStatement node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 1731
   at Microsoft.CodeAnalysis.CSharp.BoundReturnStatement.Accept(BoundTreeVisitor visitor) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 2961
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 149
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitWithStackGuard(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 359
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitAlways(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 343
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 318
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitStatement(BoundStatement statement) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 660
   at Microsoft.CodeAnalysis.CSharp.DefiniteAssignmentPass.VisitStatementsWithLocalFunctions(BoundBlock block) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\DefiniteAssignment.cs:line 1556
   at Microsoft.CodeAnalysis.CSharp.DefiniteAssignmentPass.VisitBlock(BoundBlock node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\DefiniteAssignment.cs:line 1499
   at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(BoundTreeVisitor visitor) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 2693
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 149
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitWithStackGuard(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 359
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.VisitAlways(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 343
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.Visit(BoundNode node) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 318
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.Scan(Boolean& badRegion) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 412
   at Microsoft.CodeAnalysis.CSharp.DefiniteAssignmentPass.Scan(Boolean& badRegion) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\DefiniteAssignment.cs:line 235
   at Microsoft.CodeAnalysis.CSharp.AbstractFlowPass`2.Analyze(Boolean& badRegion, Optional`1 initialState) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\AbstractFlowPass.cs:line 435
   at Microsoft.CodeAnalysis.CSharp.DefiniteAssignmentPass.Analyze(Boolean& badRegion, DiagnosticBag diagnostics) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\DefiniteAssignment.cs:line 424
   at Microsoft.CodeAnalysis.CSharp.DefiniteAssignmentPass.Analyze(CSharpCompilation compilation, MethodSymbol member, BoundNode node, DiagnosticBag diagnostics, Boolean requireOutParamsAssigned) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\DefiniteAssignment.cs:line 395
   at Microsoft.CodeAnalysis.CSharp.FlowAnalysisPass.Analyze(CSharpCompilation compilation, MethodSymbol method, BoundBlock block, DiagnosticBag diagnostics) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\FlowAnalysisPass.cs:line 135
   at Microsoft.CodeAnalysis.CSharp.FlowAnalysisPass.Rewrite(MethodSymbol method, BoundBlock block, DiagnosticBag diagnostics, Boolean hasTrailingExpression, Boolean originalBodyNested) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\FlowAnalysisPass.cs:line 49
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 1058
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 511
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass22_0.<CompileNamedTypeAsync>b__0() in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 401
   at Roslyn.Utilities.UICultureUtilities.<>c__DisplayClass5_0.<WithCurrentUICulture>b__0() in C:\Users\rikki\src\roslyn\src\Compilers\Core\Portable\InternalUtilities\UICultureUtilities.cs:line 141
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.WaitForWorkers() in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 316
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethodBodies(CSharpCompilation compilation, PEModuleBuilder moduleBeingBuiltOpt, Boolean emittingPdb, Boolean emitTestCoverageData, Boolean hasDeclarationErrors, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 155
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileMethods(CommonPEModuleBuilder moduleBuilder, Boolean emittingPdb, Boolean emitMetadataOnly, Boolean emitTestCoverageData, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in C:\Users\rikki\src\roslyn\src\Compilers\CSharp\Portable\Compilation\CSharpCompilation.cs:line 2795
   at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream metadataPEStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, CompilationTestData testData, CancellationToken cancellationToken) in C:\Users\rikki\src\roslyn\src\Compilers\Core\Portable\Compilation\Compilation.cs:line 2502
   at Roslyn.Test.Utilities.RuntimeEnvironmentUtilities.EmitCompilationCore(Compilation compilation, IEnumerable`1 manifestResources, DiagnosticBag diagnostics, CompilationTestData testData, EmitOptions emitOptions) in C:\Users\rikki\src\roslyn\src\Test\Utilities\Portable\Compilation\IRuntimeEnvironment.cs:line 252
   at Roslyn.Test.Utilities.RuntimeEnvironmentUtilities.EmitCompilation(Compilation compilation, IEnumerable`1 manifestResources, List`1 dependencies, DiagnosticBag diagnostics, CompilationTestData testData, EmitOptions emitOptions) in C:\Users\rikki\src\roslyn\src\Test\Utilities\Portable\Compilation\IRuntimeEnvironment.cs:line 225
   at Roslyn.Test.Utilities.CoreClr.CoreCLRRuntimeEnvironment.Emit(Compilation mainCompilation, IEnumerable`1 manifestResources, EmitOptions emitOptions, Boolean usePdbForDebugging) in C:\Users\rikki\src\roslyn\src\Test\Utilities\Portable\Platform\CoreClr\CoreCLRRuntimeEnvironment.cs:line 43
   at Microsoft.CodeAnalysis.Test.Utilities.CompilationVerifier.Emit(IRuntimeEnvironment testEnvironment, IEnumerable`1 manifestResources, EmitOptions emitOptions) in C:\Users\rikki\src\roslyn\src\Test\Utilities\Portable\CompilationVerifier.cs:line 235
   at Microsoft.CodeAnalysis.Test.Utilities.CompilationVerifier.Emit(String expectedOutput, Nullable`1 expectedReturnCode, String[] args, IEnumerable`1 manifestResources, EmitOptions emitOptions, Verification peVerify, SignatureDescription[] expectedSignatures) in C:\Users\rikki\src\roslyn\src\Test\Utilities\Portable\CompilationVerifier.cs:line 201
   at Microsoft.CodeAnalysis.Test.Utilities.CommonTestBase.Emit(Compilation compilation, IEnumerable`1 dependencies, IEnumerable`1 manifestResources, SignatureDescription[] expectedSignatures, String expectedOutput, Nullable`1 expectedReturnCode, String[] args, Action`1 assemblyValidator, Action`1 symbolValidator, EmitOptions emitOptions, Verification verify) in C:\Users\rikki\src\roslyn\src\Test\Utilities\Portable\CommonTestBase.cs:line 150
   at Microsoft.CodeAnalysis.Test.Utilities.CommonTestBase.CompileAndVerifyCommon(Compilation compilation, IEnumerable`1 manifestResources, IEnumerable`1 dependencies, Action`1 sourceSymbolValidator, Action`1 assemblyValidator, Action`1 symbolValidator, SignatureDescription[] expectedSignatures, String expectedOutput, Nullable`1 expectedReturnCode, String[] args, EmitOptions emitOptions, Verification verify) in C:\Users\rikki\src\roslyn\src\Test\Utilities\Portable\CommonTestBase.cs:line 64
   at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.CompileAndVerify(Compilation compilation, IEnumerable`1 manifestResources, IEnumerable`1 dependencies, Action`1 sourceSymbolValidator, Action`1 validator, Action`1 symbolValidator, SignatureDescription[] expectedSignatures, String expectedOutput, Nullable`1 expectedReturnCode, String[] args, EmitOptions emitOptions, Verification verify) in C:\Users\rikki\src\roslyn\src\Compilers\Test\Utilities\CSharp\CSharpTestBase.cs:line 816
   at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.CompileAndVerify(CSharpTestSource source, IEnumerable`1 references, IEnumerable`1 manifestResources, IEnumerable`1 dependencies, Action`1 sourceSymbolValidator, Action`1 assemblyValidator, Action`1 symbolValidator, SignatureDescription[] expectedSignatures, String expectedOutput, Nullable`1 expectedReturnCode, String[] args, CSharpCompilationOptions options, CSharpParseOptions parseOptions, EmitOptions emitOptions, TargetFramework targetFramework, Verification verify) in C:\Users\rikki\src\roslyn\src\Compilers\Test\Utilities\CSharp\CSharpTestBase.cs:line 775
```

### PR DESCRIPTION
**Pull request by RikkiGibson**
Fixes #44153

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs [hunk 1]
```diff
@@ -1953,7 +1956,7 @@ private void MarkFieldsUsed(TypeSymbol type)
 
                     var namedType = (NamedTypeSymbol)type;
                     var assembly = type.ContainingAssembly as SourceAssemblySymbol;
-                    if ((object)assembly == null)
+                    if ((object?)assembly == null)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1959]: Could simplify to 

```cs
if (type.ContainingAssembly is SourceAssemblySymbol assembly)
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs [hunk 2]
```diff
@@ -990,6 +990,19 @@ void M()
...
         }
 
+        [Fact, WorkItem(44153, "https://github.com/dotnet/roslyn/issues/44153")]
+        public void RefErrorProperty()
+        {
+            CreateCompilation(@"
+public class C {
+    public ref ERROR Prop => throw null!;
+}
+").VerifyEmitDiagnostics(
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs:1000]: >VerifyEmitDiagnostics [](start = 3, length = 21)

Does this hit the failure? From the issue description, it's not clear.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs:1000]: I made a mistake in the issue description. It's now fixed. Thanks

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs:1000]: Confirmed this fails without the change to DefiniteAssignment.

---
In reply to: [424042228](https://github.com/dotnet/roslyn/pull/44186#discussion_r424042228) [](ancestors = 424042228)

**Comment by RikkiGibson** [comment]
This is actually a classic "NRT would have caught this" bug. I'm thinking of making another PR to just `#nullable enable` in this file straight up.

**Pull request merged**